src/search/Search.java

```diff
@@ -0,0 +1,64 @@
+package search;
+
+import product.Product;
+import stock.Warehouse;
+
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * Classe per eseguire la ricerca all'interno di una Warehouse utilizzando un filtro specificato.
+ * Il filtro viene costruito utilizzando un'istanza di Filter costruita tramite FilterBuilder.
+ * Contiene una Map<Product, Integer> con i risultati della ricerca accessibile tramite .getResults();
+ */
+public class Search {
+    private final Filter filter;
+    private Map<Product, Integer> results;
+
+    /**
+     * Esegue una ricerca all'interno di una Warehouse utilizzando il filtro specificato.
+     *
+     * @param warehouse Il magazzino in cui eseguire la ricerca.
+     * @param filter    Il filtro da applicare alla ricerca, costruito utilizzando FilterBuilder.
+     */
+    public Search(Warehouse warehouse, FilterBuilder filter) {
+        this.filter = filter.createFilter();
+        results = warehouse.getStock();
+        filtering();
+    }
+
+    /**
+     * Applica i criteri di filtro ai risultati della ricerca.
+     */
+    private void filtering() {
+        filterResults(product -> filter.getArticle() == null || product.getArticle().equals(filter.getArticle()));
+        filterResults(product -> filter.getId() == 0.0 || product.getId().equals(filter.getId()));
+        filterResults(product -> filter.getProducer() == null || product.getProducer().equals(filter.getProducer()));
+        filterResults(product -> filter.getModel() == null || product.getModel().equals(filter.getModel()));
+        filterResults(product -> product.getSellPrice() >= filter.getMinSellPrice() && product.getSellPrice() <= filter.getMaxSellPrice());
+        filterResults(product -> product.getPurchasePrice() >= filter.getMinPurchasePrice() && product.getPurchasePrice() <= filter.getMaxPurchasePrice());
+        filterResults(product -> product.getDisplay() >= filter.getMinDisplay() && product.getDisplay() <= filter.getMaxDisplay());
+        filterResults(product -> product.getStorage() <= filter.getMinStorage() && product.getStorage() >= filter.getMaxStorage());
+    }
+
+    /**
+     * Applica un predicato al risultato della ricerca per filtrare i risultati.
+     *
+     * @param predicate Il preidcato da applicare al risultato della ricerca.
+     */
+    private void filterResults(Predicate<Product> predicate) {
```

Comment:
Buona l'implementazione con lo stream programming, però va rivista perché la classe in se diventa troppo complessa se utilizza un'unica classe Filter